### DISCUSSION TITLE
Short-circuiting nullable operator `?;`

### DISCUSSION THREAD

**Discussion by bert2** [dotnet/csharplang#4337] [Language Ideas] [answered]
I'd like to propose the new operator `?;` to short-circuit (i.e. conditionally early-exit) functions that return nullable types.

`?;` terminates a statement like `;` but immediately returns null from the enclosing function when the expression on its left side evaluates to `null`. The goal is to reduce boiler-plate `null`-checks like `if (x is null) return null`.

## Usage

```csharp
string? PrintUserFavorite(int userId) {
    User user = FindUser(userId)?;
    Favorite fav = user.OrderedFavorites.FirstOrDefault()?;
    return $"{fav.Name}: {fav.Description}";
}

User? FindUser(int userId) { }
```

In the above example `?;` would immediately return `null` from `PrintUserFavorite()` when `FindUser()` returns `null`. This also means that the `user` variable is guaranteed to be not `null`. The same is true for the call to `FirstOrDefault()` and the variable `fav`.

The compiler would emit a warning when `?;` is used inside a function that doesn't return a nullable type.

## Possible implementation

`var x = y?;` would desugar to something like `var x = y; if (x is null) return null;`

The example from above, but desugared:

```csharp
string? PrintUserFavorite(int userId) {
    User? user = FindUser(userId);
    if (user is null) return null;

    Favorite? fav = user.OrderedFavorites.FirstOrDefault();
    if (fav is null) return null;

    return $"{fav.Name}: {fav.Description}";
}
```

### Use `?.`

The safe navigation operator `?.` already gets us quite far in terms of short-circuiting behavior, but can only avoid `null` checks when the function's logic is written as a single expression and does not require multiple statements.

The above example using `?.` where possible:

```csharp
string? PrintUserFavorite(int userId) {
    Favorite? fav = FindUser(userId)?.OrderedFavorites.FirstOrDefault();
    return fav is not null ? $"{fav.Name}: {fav.Description}" : null;
}
```

### Allow `return` on right-hand side of `??`

My favorite alternative, because it does not introduce a new operator, is more flexible than `?;` (e.g. `... ?? return "no favorite"`), and conceptually aligns with `throw` expressions. However, this has been [discussed](https://github.com/dotnet/csharplang/issues/867) already, and it doesn't look like it will ever be implemented.

```csharp
string? PrintUserFavorite(int userId) {
    User user = FindUser(userId) ?? return null;
    // ...
```

### Make `T?` a functor/monad

Given the [right extension methods](https://github.com/bert2/Nullable.Extensions), the `PrintUserFavorite()` example could be rewritten to:

```csharp
string? PrintUserFavorite(int userId) =>
    FindUser(userId)?
    .OrderedFavorites
    .FirstOrDefault()
    .Map(fav => $"{fav.Name}: {fav.Description}");
```

C# programmers unfamiliar with concepts from functional programming might find this rather off-putting. Also the implementation of those extension methods is a quite messy, because they have to handle both nullable value types and nullable reference types.

### The null-safe navigation operator `?.`

IMHO `?;` feels and looks like the natural extension of `?.` to assignment statements and both work well together:

```csharp
string? PrintUserFavorite(int userId) {
    Favorite fav = FindUser(userId)?.OrderedFavorites.FirstOrDefault()?;
    return $"{fav.Name}: {fav.Description}";
}
```

### Rust's `Try` trait

Rust has a trait that enables the similar [`?` operator](https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html#a-shortcut-for-propagating-errors-the--operator) on all types that implement it. The standard library implements it for `Result<T, E>` and `Option<T>` which are the same as F#'s `Result<'T, 'TError>` and `Option<'T>`.

Here's an example of its usage:

```rust
fn read_username_from_file() -> Result<String, io::Error> {
    let mut f = File::open("hello.txt")?; // early-exits in case there was an `io::Error`
    let mut s = String::new();
    f.read_to_string(&mut s)?; // early-exits in case there was an `io::Error`
    Ok(s)
}
```

Rust's `?` operator also works inside expressions, e.g. `foo()?.bar()?.baz()?`.

**Comment by svick** [comment]
My problem with this suggestion is that it makes it harder to understand what the code does, because it's just a single symbol at the end of the line, which completely changes its control flow. It's just way too subtle for what it does.

**Reply by bert2** [reply]
I had the exact same thoughts when I started a project with Rust, but that changed quickly after using it a couple times. It's just so much more ergonomic.

And `?.` is not so much different in terms of changes to the control flow when you consider methods with expression bodies:

```csharp
static string? GetFirstName(this IEnumerable<User> users) => users.FirstOrDefault()?.Name;
```

The subtle `?` in the middle of the expression also completely changes the control flow in very much the same way, but I doubt anybody would argue against `?.` for that reason.

**Reply by 333fred** [reply]
Rust has several of these types of single-character changes at the end of statements, such as `;` being the equivalent of F#'s `|> ignore`. While they work well there, I just don't think they work well in C#, particularly given that `?.` already means something and there are no other implied early-return operators. I would find it highly confusing, personally.

**Comment by CyrusNajmabadi** [comment]
I very much like what rust does here.  But htey also have the benefit of being a new language and installing this sort of design very early in their design.  That doesn't mean we can't add this ourselves.  But it does mean we have to better understand what impact it would have in the context of millions of devs who are used to how C# has generally done things for the last 20 years.

**Comment by spydacarnage** [comment, marked as answer]
I think I'd prefer the idea of `string name = GetName() ?? return "";` over the `?;` idea - that's too easy to miss, and the return idea follows the throw pattern we already have.

**Reply by HaloFour** [reply]
Agreed, I'd rather revisit `return` expressions.

**Reply by CyrusNajmabadi** [reply]
I'm willing to champion. Someone linky me?

**Reply by HaloFour** [reply]
> Update: Let's consider again when we have match expressions, where they may become more useful.

Take 2?  😉 

https://github.com/dotnet/csharplang/issues/867

**Reply by 333fred** [reply]
I see us discussing this type of thing when we consider block expressions.

**Reply by bert2** [reply]
> Take 2? 😉

Plot twist: that was my plan all along! 😄 

![](https://media3.giphy.com/media/xl5QdxfNonh3q/giphy.gif)

**Reply by orthoxerox** [reply]
@CyrusNajmabadi @333fred why not champion user-defined `?` semantics along with this (to allow `?.` and `??` on `Option<T>`)? In for a penny, in for a pound!
